### PR TITLE
Add easy component updating functionality

### PR DESCRIPTION
**Pull request by WulffHunter**
Currently, updating the contents of a component is a largely manual process, where one must make sure they remember all the values that they need to update in their component. For large components, this can be an issue.

This PR adds functionality to take data in a similar structure to the shape of the component and unpack it into an entity's component. It also adds strong typing to ensure that, when used with TypeScript, the data shape will match that of the component.

**Currently, tests are failing as `$entitySparseSet` cannot be found in `addComponent`**. This does not seem to be an issue introduced by this PR. However, until then, tests are failing for this functionality.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -101,9 +102,14 @@ world.time = { delta: 0, elapsed: 0, then: performance.now() }
...
-addComponent(world, Velocity, eid)
-Velocity.x[eid] = 1.23
-Velocity.y[eid] = 1.23
+Position.x[eid] = 1.23
+Position.y[eid] = 1.23
+// An alternative way of adding and filling components:
+addAndFillComponent(world, Velocity, eid, {
+  x: 4.56,
+  y: 4.56,
+  x: 0,
```
Comment (vegeta897) [inline README.md:111]: This duplicate `x` property seems unintentional

Comment (WulffHunter) [reply README.md:111]: Good point! That's supposed to be a `z`. Thanks!

**Comment by WulffHunter** [comment]
Currently I'm seeing two failures:
```
  1) Component Integration Tests
       should add and fill, and update components from an entity:
     TypeError: Cannot read properties of undefined (reading 'has')
      at addComponent (file://~Documents/bitECS/dist/index.mjs:957:32)
      at addAndPartiallyFillComponent (file://~/Documents/bitECS/dist/index.mjs:992:3)
      at addAndFillComponent (file://~/Documents/bitECS/dist/index.mjs:996:3)
      at Context.<anonymous> (file://~/Documents/bitECS/test/integration/Component.test.js:55:5)
      at process.processImmediate (node:internal/timers:471:21)

  2) Entity Integration Tests
       should flush entity IDs:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

0 !== 1500

      + expected - actual

      -0
      +1500
      
      at Context.<anonymous> (file://~/Documents/bitECS/test/integration/Entity.test.js:97:5)
      at process.processImmediate (node:internal/timers:471:21)
```
The first one fails as `world[$entitySparseSet]` is `undefined`. However, in the test function, I've confirmed that this is not the case: it's only the case once we're inside of the `addAndFillComponent()` function. @NateTheGreatt have you encountered this behaviour in your tests before?

I'm not sure why the second test is failing. I haven't modified anything in it.

**Comment by NateTheGreatt** [comment]
Really appreciate this. One of the most substantial PRs I've seen yet, and this has been a frequent API request.

Currently, @SupremeTechnopriest is in the middle of a full rewrite of bitECS in Typescript, with a some new features and additional goodies. So unfortunately this PR will need to be rewritten for that rewrite 😅. 

Something I want to discuss before adding functions like this to the API is that I want to keep bitECS as GC-less as possible. For performance reasons, I avoid JS object creation in API functions at all costs. Objects are only created when absolutely necessary. This PR currently includes functions which create objects each call. The garbage collector will eventually need to take care of those objects. It may seem like a small amount of garbage, until you start scaling up and calling these functions many times in batches (like bullet hell games or MMOs).

While the functions are useful, including them in the API might encourage people to use them whether or not they fully understand the performance implications, and it would also prevent people from asking why these functions aren't included by default.

All that said, I'd like to keep the discussion open on this until the rewrite is done (it's super close). Maybe these new API functions just need a performance warning in the docs?

**Comment by NateTheGreatt** [comment]
there are some mechanisms for this by using observers in the new version:

https://github.com/NateTheGreatt/bitECS/tree/rc-0-4-0/docs/Intro.md

**Comment by WulffHunter** [comment]
It's been a few years since I've used this repo, I'm super excited to see that the new version includes this functionality. It looks fantastic! Very excited to use it!

**Pull request closed without merging**